executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -65,200 +63,41 @@ impl<'a, T: FieldElement> BlockMachineProcessor<'a, T> {
-            }
-        }
+        let identities = self.row_range().flat_map(move |row| {
+            self.machine_parts
+                .identities
+                .iter()
+                .map(move |&id| (id, row))
+        });
+        let requested_known = known_args
+            .iter()
+            .enumerate()
+            .filter_map(|(i, is_input)| (!is_input).then_some(Variable::Param(i)));
+        Processor::new(
+            self.fixed_data,
+            self,
+            identities,
+            self.block_size,
+            requested_known,
+        )
+        .generate_code(can_process, witgen)
+        .map_err(|e| {
+            log::debug!("\nCode generation failed for connection:\n  {connection}");
+            let known_args_str = known_args
+                .iter()
+                .enumerate()
+                .filter_map(|(i, b)| b.then_some(connection.right.expressions[i].to_string()))
+                .join("\n  ");
+            log::debug!("Known arguments:\n  {known_args_str}");
+            log::debug!("Error:\n  {e}");
+            format!("Code generation failed: {e}\nRun with RUST_LOG=debug to see the code generated so far.")
+        })
     }
 
     fn row_range(&self) -> std::ops::Range<i32> {
         // We iterate over all rows of the block +/- one row, so that we can also solve for non-rectangular blocks.
         -1..(self.block_size + 1) as i32
     }
-
-    /// Repeatedly processes all identities on all rows, until no progress is made.
-    /// Fails iff there are incomplete machine calls in the latch row.
-    fn solve_block<CanProcess: CanProcessCall<T> + Clone>(
-        &self,
-        can_process: CanProcess,
-        witgen: &mut WitgenInference<'a, T, &Self>,
-        connection_rhs: &SelectedExpressions<T>,
-    ) -> Result<(), String> {
-        let mut complete = HashSet::new();
-        for iteration in 0.. {
-            let mut progress = false;
-
-            for row in self.row_range() {
-                for id in &self.machine_parts.identities {
-                    if !complete.contains(&(id.id(), row)) {
-                        let result = witgen.process_identity(can_process.clone(), id, row);
-                        if result.complete {
-                            complete.insert((id.id(), row));
-                        }
-                        progress |= result.progress;
-                    }
-                }
-            }
-            if !progress {
-                log::trace!(
-                    "Finishing block machine witgen code generation after {iteration} iterations"
-                );
-                break;
-            }
-        }
-
-        for (index, expr) in connection_rhs.expressions.iter().enumerate() {
-            if !witgen.is_known(&Variable::Param(index)) {
-                return Err(format!(
-                    "Unable to derive algorithm to compute output value \"{expr}\""
```

Comment:
This error message cannot be reproduced currently, it only says `param[0]` instead of the proper expression.